p2p/module.go

```diff
@@ -123,31 +153,54 @@ func (m *p2pModule) Start() error {
 			telemetry.P2P_NODE_STARTED_TIMESERIES_METRIC_DESCRIPTION,
 		)
 
-	go func() {
-		for {
-			data, err := m.listener.ReadAll()
-			if err != nil {
-				m.logger.Error().Err(err).Msg("Error reading data from connection")
-				continue
-			}
-			go m.handleNetworkMessage(data)
-		}
-	}()
+	if err = m.startHost(); err != nil {
+		return fmt.Errorf("starting libp2pHost: %w", err)
+	}
+
+	listenAddrLogEvent := m.logger.Info()
+	for i, addr := range libp2pHost.InfoFromHost(m.host).Addrs {
+		listenAddrLogEvent.Str(fmt.Sprintf("listen_addr_%d", i), addr.String())
+	}
+	listenAddrLogEvent.Msg("Listening for incoming connections...")
+
+	// TECHDEBT: use RandomSub or GossipSub once we're on more stable ground.
+	// IMPROVE: consider supporting multiple router types via config.
+	m.pubsub, err = pubsub.NewFloodSub(ctx, m.host)
+	if err != nil {
+		return fmt.Errorf("unable to create pubsub: %w", err)
+	}
+
+	// Topic is used to `#Publish` messages.
+	m.topic, err = m.pubsub.Join(protocol.DefaultTopicStr)
+	if err != nil {
+		return fmt.Errorf("unable to join pubsub topic: %w", err)
+	}
+
+	// Subscription is notified when a new message is received on the topic.
+	m.subscription, err = m.topic.Subscribe()
+	if err != nil {
+		return fmt.Errorf("subscribing to pubsub topic: %w", err)
+	}
+
+	if err := m.startNetwork(); err != nil {
+		return fmt.Errorf("creating network: %w", err)
+	}
+
+	// Don't handle streams or read from the subscription in client debug mode.
+	if !m.isClientDebugMode() {
+		m.host.SetStreamHandler(protocol.PoktProtocolID, m.handleStream)
+		go m.readFromSubscription(ctx)
+	}
 
 	m.GetBus().
 		GetTelemetryModule().
 		GetTimeSeriesAgent().
 		CounterIncrement(telemetry.P2P_NODE_STARTED_TIMESERIES_METRIC_NAME)
-
 	return nil
 }
 
 func (m *p2pModule) Stop() error {
-	m.logger.Info().Msg("Stopping network module")
-	if err := m.listener.Close(); err != nil {
-		return err
-	}
-	return nil
+	return m.host.Close()
```

Comment:
1. Yes - nice catch! :smile: However, as we're apparently not actually using pubsub at this time, I get to be lazy on this one. :sweat_smile: 
2. No, which is why I'm proposing to rename it to `setupNetwork` instead.

Now that I'm looking at it in context, I think we should also set `m.host` to `nil` in `#Stop()` so that `#Start()` will re-create it rather than try to use a closed host.